src/main/java/com/adamlewis/guice/persist/jooq/JooqPersistService.java

```diff
@@ -90,7 +90,11 @@ public void begin() {
     DefaultConnectionProvider conn;
     try {
       logger.debug("Getting JDBC connection");
-      conn = new DefaultConnectionProvider(jdbcSource.getConnection());
+      DataSource dataSource = jdbcSource.get();
+      if (dataSource == null) {
+        throw new RuntimeException("Datasource not available from provider");
+      }
+      conn = new DefaultConnectionProvider(jdbcSource.get().getConnection());
```

Comment:
This code calls the Provider twice.  Once for the null check and then again in the `DefaultConnectionProvider` constructor.  I don't think the `null` check is needed since Guice shouldn't inject `null`.  And if it did, the default NPE would be more specific anyway.
```suggestion
      DataSource dataSource = jdbcSource.get();
      Connection jdbcConn = dataSource.getConnection();
      conn = new DefaultConnectionProvider(jdbcConn);
```

Reply:
Ah, good catch on the duplicate calls to the `Provider.get()` method; I'll fix that.

I do think it's still possible that an injected `Provider` _could_ return a `null` datasource. It's bad form to do so, but this is just a little bit of defensive coding to help indicate where things broke down.

Reply:
Yeah, got that...Guice docs say it shouldn't happen in most cases, but then backtrack that a bit, reference an old oversight, etc...so it _could_ happen.  In any case, I split the series of dereferences onto their own lines.  If `dataSource` is `null` then the VM will raise a `NullPointerException` on line 94 which should unambiguously indicate what went wrong and why.